ICE: `no entry found for key`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' ($TID) panicked at compiler/rustc_resolve/src/diagnostics.rs:2183:33: 'no entry found for key'', 'thread 'rustc' ($TID) panicked at compiler/rustc_resolve/src/diagnostics.rs:2183:33: 'no entry found for key''
File: /tmp/im/a.rs
-->

snippet:
````rust
mod m {
    struct Priv2;
}
fn main() {
    WithUse { one: m::Priv2 }
}

````

Version information
````
rustc 1.91.0-nightly (350d0ef0e 2025-08-13)
binary: rustc
commit-hash: 350d0ef0ec0493e6d21cfb265cb8211a0e74d766
commit-date: 2025-08-13
host: x86_64-unknown-linux-gnu
release: 1.91.0-nightly
LLVM version: 21.1.0

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/350d0ef0ec0493e6d21cfb265cb8211a0e74d766/compiler/rustc_resolve/src/diagnostics.rs#L2177-L2189

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0422]: cannot find struct, variant or union type `WithUse` in this scope
 --> /tmp/icemaker_global_tempdir.hgrQWhVcL346/rustc_testrunner_tmpdir_reporting.We18CgZLUs4Y/mvce.rs:5:5
  |
5 |     WithUse { one: m::Priv2 }
  |     ^^^^^^^ not found in this scope


thread 'rustc' (2557190) panicked at compiler/rustc_resolve/src/diagnostics.rs:2183:33:
no entry found for key
stack backtrace:
   0:     0x7f7140de5a43 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hff1a9408d012ee89
   1:     0x7f7141601bfc - core::fmt::write::h2f1fc679f228cc52
   2:     0x7f7140dd0493 - std::io::Write::write_fmt::haa2231cbd9aebf95
   3:     0x7f7140de58a2 - std::sys::backtrace::BacktraceLock::print::h28bfdbcac6d3828b
   4:     0x7f7140dc7c1f - std::panicking::default_hook::{{closure}}::h1f968e3a24b0e3b4
   5:     0x7f7140dc774b - std::panicking::default_hook::h43da101644971931
   6:     0x7f713fe491b5 - std[619c96d97fe952c]::panicking::update_hook::<alloc[a5c46f2b217994b5]::boxed::Box<rustc_driver_impl[c8262843f73b2e4e]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x7f7140dc8313 - std::panicking::panic_with_hook::hb4e916c3e65eccec
   8:     0x7f7140de544a - std::panicking::panic_handler::{{closure}}::hbff715d0475cf1bd
   9:     0x7f7140de53a9 - std::sys::backtrace::__rust_end_short_backtrace::hd6f43484e45a4955
  10:     0x7f7140dc89ed - __rustc[b29d3319de9a3146]::rust_begin_unwind
  11:     0x7f713d4d8890 - core::panicking::panic_fmt::h5f8d281875be5ff1
  12:     0x7f713e5837db - core::option::expect_failed::h7c4e9e1c4059c90a
  13:     0x7f714097278f - <rustc_resolve[c941e31c2824c6c5]::Resolver>::report_privacy_error
  14:     0x7f7141f4535b - <rustc_resolve[c941e31c2824c6c5]::Resolver>::resolve_crate::{closure#0}
  15:     0x7f714217af7f - <rustc_resolve[c941e31c2824c6c5]::Resolver>::resolve_crate
  16:     0x7f7142179269 - rustc_interface[b77c32c3bfe8d9]::passes::configure_and_expand
  17:     0x7f71428b4b2d - rustc_interface[b77c32c3bfe8d9]::passes::resolver_for_lowering_raw
  18:     0x7f71428b488d - rustc_query_impl[e2f117ce3af55616]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2f117ce3af55616]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[5ca2e9449305faae]::query::erase::Erased<[u8; 16usize]>>
  19:     0x7f71428b4867 - <rustc_query_impl[e2f117ce3af55616]::query_impl::resolver_for_lowering_raw::dynamic_query::{closure#2} as core[b7f6e7468eb674bd]::ops::function::FnOnce<(rustc_middle[5ca2e9449305faae]::ty::context::TyCtxt, ())>>::call_once
  20:     0x7f71426a765f - rustc_query_system[c907f34f82b7658f]::query::plumbing::try_execute_query::<rustc_query_impl[e2f117ce3af55616]::DynamicConfig<rustc_query_system[c907f34f82b7658f]::query::caches::SingleCache<rustc_middle[5ca2e9449305faae]::query::erase::Erased<[u8; 16usize]>>, false, false, false>, rustc_query_impl[e2f117ce3af55616]::plumbing::QueryCtxt, false>
  21:     0x7f71426a71bc - rustc_query_impl[e2f117ce3af55616]::query_impl::resolver_for_lowering_raw::get_query_non_incr::__rust_end_short_backtrace
  22:     0x7f714278ecb6 - rustc_interface[b77c32c3bfe8d9]::passes::create_and_enter_global_ctxt::<core[b7f6e7468eb674bd]::option::Option<rustc_interface[b77c32c3bfe8d9]::queries::Linker>, rustc_driver_impl[c8262843f73b2e4e]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  23:     0x7f71427efe03 - rustc_interface[b77c32c3bfe8d9]::interface::run_compiler::<(), rustc_driver_impl[c8262843f73b2e4e]::run_compiler::{closure#0}>::{closure#1}
  24:     0x7f714274e7fb - std[619c96d97fe952c]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[b77c32c3bfe8d9]::util::run_in_thread_with_globals<rustc_interface[b77c32c3bfe8d9]::util::run_in_thread_pool_with_globals<rustc_interface[b77c32c3bfe8d9]::interface::run_compiler<(), rustc_driver_impl[c8262843f73b2e4e]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  25:     0x7f714274e4d4 - <<std[619c96d97fe952c]::thread::Builder>::spawn_unchecked_<rustc_interface[b77c32c3bfe8d9]::util::run_in_thread_with_globals<rustc_interface[b77c32c3bfe8d9]::util::run_in_thread_pool_with_globals<rustc_interface[b77c32c3bfe8d9]::interface::run_compiler<(), rustc_driver_impl[c8262843f73b2e4e]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[b7f6e7468eb674bd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  26:     0x7f714274a28b - std::sys::pal::unix::thread::Thread::new::thread_start::h66ff1bba185a50cf
  27:     0x7f713c0a57eb - <unknown>
  28:     0x7f713c12918c - <unknown>
  29:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.91.0-nightly (350d0ef0e 2025-08-13) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [resolver_for_lowering_raw] getting the resolver for lowering
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0422`.

```

</p>
</details>

<!--
query stack:
#0 [resolver_for_lowering_raw] getting the resolver for lowering

-->

--- a/compiler/rustc_resolve/src/diagnostics.rs
+++ b/compiler/rustc_resolve/src/diagnostics.rs
@@ -2189,9 +2189,9 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
         let ast::ExprKind::Struct(struct_expr) = &expr.kind else { return };
         // We don't have to handle type-relative paths because they're forbidden in ADT
         // expressions, but that would change with `#[feature(more_qualified_paths)]`.
-        let Some(Res::Def(_, def_id)) =
-            self.partial_res_map[&struct_expr.path.segments.iter().last().unwrap().id].full_res()
-        else {
+        let Some(segment) = struct_expr.path.segments.last() else { return };
+        let Some(partial_res) = self.partial_res_map.get(&segment.id) else { return };
+        let Some(Res::Def(_, def_id)) = partial_res.full_res() else {
             return;
         };
         let Some(default_fields) = self.field_defaults(def_id) else { return };

--- a/tests/ui/structs/default-field-values/non-exhaustive-ctor-not-found.rs
+++ b/tests/ui/structs/default-field-values/non-exhaustive-ctor-not-found.rs
@@ -1,0 +1,8 @@
+// Regression test for https://github.com/rust-lang/rust/issues/145367
+mod m {
+    struct Priv2;
+}
+fn main() {
+    WithUse { one: m::Priv2 } //~ ERROR: cannot find struct, variant or union type `WithUse` in this scope
+    //~^ ERROR: unit struct `Priv2` is private
+}

--- a/tests/ui/structs/default-field-values/non-exhaustive-ctor-not-found.stderr
+++ b/tests/ui/structs/default-field-values/non-exhaustive-ctor-not-found.stderr
@@ -1,0 +1,22 @@
+error[E0422]: cannot find struct, variant or union type `WithUse` in this scope
+  --> $DIR/non-exhaustive-ctor-not-found.rs:6:5
+   |
+LL |     WithUse { one: m::Priv2 }
+   |     ^^^^^^^ not found in this scope
+
+error[E0603]: unit struct `Priv2` is private
+  --> $DIR/non-exhaustive-ctor-not-found.rs:6:23
+   |
+LL |     WithUse { one: m::Priv2 }
+   |                       ^^^^^ private unit struct
+   |
+note: the unit struct `Priv2` is defined here
+  --> $DIR/non-exhaustive-ctor-not-found.rs:3:5
+   |
+LL |     struct Priv2;
+   |     ^^^^^^^^^^^^^
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0422, E0603.
+For more information about an error, try `rustc --explain E0422`.

